File: output/ChatGPT/avatar/Python/Java/atcoder_ABC165_C.java
import java.util.*;

public class atcoder_ABC165_C {
    static int res;

    public static void solve(int N, int M, int Q, int[] a, int[] b, int[] c, int[] d) {
        res = 0;
        rec(new ArrayList<>(), N, M, Q, a, b, c, d);
        System.out.println(res);
    }

    public static void rec(List<Integer> A, int N, int M, int Q, int[] a, int[] b, int[] c, int[] d) {
        if (A.size() == N) {
            int ans = 0;
            for (int i = 0; i < Q; i++) {
                if (A.get(b[i] - 1) - A.get(a[i] - 1) == c[i]) {
                    ans += d[i];
                }
            }
            res = Math.max(res, ans);
            return;
        }
        int last_max = A.isEmpty() ? 1 : A.get(A.size() - 1);
        for (int i = last_max; i <= M; i++) {
            List<Integer> newA = new ArrayList<>(A);
            newA.add(i);
            rec(newA, N, M, Q, a, b, c, d);
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int N = scanner.nextInt();
        int M = scanner.nextInt();
        int Q = scanner.nextInt();
        int[] a = new int[Q];
        int[] b = new int[Q];
        int[] c = new int[Q];
        int[] d = new int[Q];
        for (int i = 0; i < Q; i++) {
            a[i] = scanner.nextInt();
            b[i] = scanner.nextInt();
            c[i] = scanner.nextInt();
            d[i] = scanner.nextInt();
        }
        solve(N, M, Q, a, b, c, d);
    }
}

File: output/ChatGPT/avatar/Python/C++/atcoder_ABC165_C.cpp
#include <iostream>
#include <vector>

using namespace std;

int res = 0;

void rec(int N, int M, int Q, vector<int>& A, vector<int>& a, vector<int>& b, vector<int>& c, vector<int>& d) {
    if (A.size() == N) {
        int ans = 0;
        for (int i = 0; i < Q; i++) {
            if (A[b[i] - 1] - A[a[i] - 1] == c[i]) {
                ans += d[i];
            }
        }
        res = max(res, ans);
        return;
    }
    int last_max = A.empty() ? 1 : A.back();
    for (int i = last_max; i <= M; i++) {
        A.push_back(i);
        rec(N, M, Q, A, a, b, c, d);
        A.pop_back();
    }
}

void solve(int N, int M, int Q, vector<int>& a, vector<int>& b, vector<int>& c, vector<int>& d) {
    res = 0;
    vector<int> A;
    rec(N, M, Q, A, a, b, c, d);
    cout << res << endl;
}

int main() {
    int N, M, Q;
    cin >> N >> M >> Q;
    vector<int> a(Q), b(Q), c(Q), d(Q);
    for (int i = 0; i < Q; i++) {
        cin >> a[i] >> b[i] >> c[i] >> d[i];
    }
    solve(N, M, Q, a, b, c, d);
    return 0;
}